libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts

```diff
@@ -0,0 +1,114 @@
+  globalExcludedGlobs: string[];
+  projects: {
+    [key: string]: IProjectConfiguration;
+  };
+}
+
+export class ProjectImpactGraphGenerator {
+  /**
+   * Full path of repository root
+   */
+  private _repositoryRoot: string;
+
+  /**
+   * Projects within the Rush configuration
+   */
+  private _projects: RushConfigurationProject[];
+
+  /**
+   * Default global excluded globs
+   * Only used if the `<repository_root>/.mergequeueignore` does not exist
+   */
+  private _DefaultGlobalExcludedGlobs: string[] = ['common/autoinstallers/**'];
+
+  /**
+   * Get repositoryRoot and load projects within the rush.json
+   */
+  public constructor(rushConfiguration: RushConfiguration) {
+    this._repositoryRoot = rushConfiguration.rushJsonFolder;
+    this._projects = rushConfiguration.projects;
+  }
+
+  /**
+   * Load global excluded globs
+   * @param repositoryRoot
+   */
+  private _loadGlobalExcludedGlobs(repositoryRoot: string): string[] | undefined {
+    const filePath: string = path.join(repositoryRoot, '.mergequeueignore');
+    if (fs.existsSync(filePath)) {
+      const globs: string[] = fs.readFileSync(filePath).toString().split('\n');
+      return globs;
+    }
+  }
+
+  /**
+   * Load project excluded globs
+   * @param projectRootRelativePath - project root relative path
+   */
+  private _loadProjectExcludedGlobs(projectRootRelativePath: string): string[] | undefined {
+    const filePath: string = path.join(this._repositoryRoot, projectRootRelativePath, '.mergequeueignore');
+    if (fs.existsSync(filePath)) {
+      const globs: string[] = fs.readFileSync(filePath).toString().split('\n');
+      return globs.map((glob) => path.join(projectRootRelativePath, glob));
+    }
+  }
+
+  /**
+   * Core Logic: generate project-impact-graph.yaml
+   */
+  public generate(): void {
+    const stopwatch: Stopwatch = Stopwatch.start();
+    const content: IFileSchema = {} as IFileSchema;
+    content.globalExcludedGlobs =
+      this._loadGlobalExcludedGlobs(this._repositoryRoot) || this._DefaultGlobalExcludedGlobs;
+    content.projects = {};
+    this._projects.forEach((project) => {
+      // ignore the top project
+      if (project.projectRelativeFolder !== '.') {
+        const dependentList: string[] = [project.packageName];
+        project.consumingProjects.forEach((item) => {
+          dependentList.push(item.packageName);
+        });
+        content.projects[`${project.packageName}`] = {
+          includedGlobs: [`${project.projectRelativeFolder}/**`],
+          dependentProjects: dependentList.sort()
+        };
+        const projectExcludedGlobs: string[] | undefined = this._loadProjectExcludedGlobs(
+          project.projectRelativeFolder
+        );
+        if (projectExcludedGlobs) {
+          content.projects[`${project.packageName}`].excludedGlobs = projectExcludedGlobs;
+        }
+      }
+    });
+
+    fs.writeFileSync(path.join(this._repositoryRoot, 'project-impact-graph.yaml'), yaml.safeDump(content));
```

Comment:
Seems like this file path should be configurable.

Reply:
```suggestion
    fs.writeFileSync(`${this._repositoryRoot}/project-impact-graph.yaml`, yaml.safeDump(content));
```

Reply:
@iclanton The YAML file is meant to be read directly from Git by external systems such as GitHub or Mergify, which are language agnostic. This is similar to other files such as `CODEOWNERS` or `.gitattributes`.  We can make it configurable, but then those integrating systems would need some way to discover or store the file path. And is it merely the location that is configurable, but the filename part is standard? What do you think?

If the location is in the repository root, then maybe the filename should be `.project-impact-graph.yaml`.

Reply:
I like that name as well.